node/src/components/consensus/highway_core/state.rs

```diff
@@ -115,6 +115,61 @@ impl<C: Context> Fault<C> {
     }
 }
 
+#[derive(Debug, Clone)]
+pub(crate) struct Panoramas<C: Context> {
+    /// The full panorama, corresponding to the complete protocol state.
+    panorama: Panorama<C>,
+    /// Panorama used when creating new units.
+    /// In the presence of faults may lag behind `panorama`.
+    /// Units, that if cited by a new unit would make that unit violate the LNC, are not added to
+    /// `citable_panorama`.
+    citable_panorama: Panorama<C>,
+}
+
+impl<C: Context> Panoramas<C> {
+    fn new(panorama: Panorama<C>, citable_panorama: Panorama<C>) -> Self {
+        Panoramas {
+            panorama,
+            citable_panorama,
+        }
+    }
+
+    /// Returns the complete protocol state's latest panorama.
+    pub(crate) fn panorama(&self) -> &Panorama<C> {
+        &self.panorama
+    }
+
+    /// Returns the citable panorama.
+    pub(crate) fn citable_panorama(&self) -> &Panorama<C> {
+        &self.citable_panorama
+    }
+
+    /// Marks validator at `idx` as faulty.
+    pub(crate) fn mark_faulty(&mut self, idx: ValidatorIndex) {
+        self.panorama[idx] = Observation::Faulty;
+        self.citable_panorama[idx] = Observation::Faulty;
+    }
+
+    /// Updates latest observation of `creator` as `Correct(unit)`.
+    /// Once a unit is a endorsed it is safe to cite it.
+    pub(crate) fn endorsed(&mut self, creator: ValidatorIndex, unit: C::Hash) {
+        self.citable_panorama[creator] = Observation::Correct(unit);
```

Comment:
We need to include everything seen by `unit` as well, otherwise this could make `citable_panorama` invalid. E.g. if `unit2` by Alice is seen from `unit` but later than `citable_panorama[ALICE]`, we _must_ set `citable_panorama[ALICE] = Correct(unit2)`, otherwise any vote with this panorama will be considered invalid.

Reply:
Also, if `unit` happens to be _less_ than `citable_panorama[creator]`, we shouldn't update it at all here.

Reply:
Right, I assumed that units will be get endorsed from oldest to the latest. Then everything seen by `unit` would get endorsed before `unit` and `unit` would never be less than `citable_panorama[creator]`.

Reply:
I don't think that's safe to assume. Endorsements could arrive in any order.

Reply:
Yes, I missed that. I will take a not and fix in the next PR.

Reply:
I'd prefer not to merge this into `master` without the fix, though, since we are already using `citable_panorama`, and now it's not guaranteed that it's always equal to `panorama` (or valid at all) because in `endorsed` we overwrite stuff.

Reply:
Remove the body of that method for now.